### ISSUE TITLE
Add SPARQL support

### ISSUE THREAD

**Issue by pchampin** [pchampin/sophia_rs#19]
- [ ] add a (set of) type(s) for representing a [SPARQL abstract query](https://www.w3.org/TR/sparql11-query/#sparqlDefinition) 
- [ ] add a parser for building a SPARQL abstract query (see above) from a string in the SPARQL syntax
- [ ] incrementally implement the SPARQL algebra (this item should probably be split into sub-items)

NB: since sophia uses a *generalized* RDF model (including variables), a [Graph](https://docs.rs/sophia/0.3.0/sophia/graph/trait.Graph.html) can also be used as a [basic graph pattern](https://www.w3.org/TR/sparql11-query/#BasicGraphPatterns). The [query](https://github.com/pchampin/sophia_rs/blob/master/sophia/src/query.rs) module contains a preliminary implementation of this idea.

**Comment by Tpt** [comment]
For Oxigraph I built a SPARQL parser and a SPARQL algebra representation.

Algebra: https://github.com/Tpt/oxigraph/blob/master/lib/src/sparql/algebra.rs
Parser: https://github.com/Tpt/oxigraph/blob/master/lib/src/sparql/sparql_grammar.rustpeg
Parser invocation: https://github.com/Tpt/oxigraph/blob/master/lib/src/sparql/parser.rs

It might be interesting to build it as a separated crate and make Sophia and Oxigraph depend on it, just like Rio.

The parser is a bit slow at the moment, I am planning to rewrite it using a more efficient parsing library, probably nom. But I plan to make a working 0.1 release of Oxigraph first.

**Comment by pchampin** [comment]
Thanks @Tpt for chiming in.

> It might be interesting to build it as a separated crate and make Sophia and Oxigraph depend on it, just like Rio.

I have considered this. But as I mentioned above, in sophia it would be more natural to reuse the `graph::Graph` to represent basic graph patterns, so this might not be the smoother way to go... I'm still open to ideas, though. The more work we can mutualize, the better.

**Comment by Tpt** [comment]
> The more work we can mutualize, the better.

Huge +1. We could maybe have the parser in a separated crate with a fairly cheap algebra representation. Then Sophia could expose an easy to use algebra tree on top of it and Oxigraph could build from it its query plans.

An other way to go would be to have an "rdf-api" crate similar to what RDF/JS is doing for the RDF models and its commons extensions. And have Oxigraph and Sophia and hopefully the other RDF related libraries in Rust use it. But it might be hard to build a nice and efficient API without [GAT](https://github.com/rust-lang/rust/issues/44265).

**Comment by pchampin** [comment]
> An other way to go would be to have an "rdf-api" crate similar to what RDF/JS

This should probably be discussed in a separate thread. I created #23 for this. And yes, GAT would be a huge help in this direction.

**Comment by dwhitney** [comment]
Isaac Newton invented calculus while quarantined, so I guess I can write a SPARQL parser? I've used the oxigraph library quite a bit, and I like it, but the slowest part of it is its parser (acknowledged by @Tpt). I guess I can just start writing one and then ask for feedback? I'd like to to parse into a common AST. I guess using the oxigraph algebra is sufficient?

**Comment by pchampin** [comment]
@dwhitney that would awesome...  :-)

Sophia has evolved quite a bit in the meantime, in order to be more usable as a common API for RDF in Rust. @Tpt and I have agreed (in a discussion offline) that a good way forward would be to extract oxigraph's SPARQL parser and AST into a separate crate, using sophia's `Term` type as a building block.

FYI, the `Term` type is currently being refactored (#47, #48, #49). Once this is finished (Literals still need to be done), I plan to extract it into a separate crate `sophia_term`, so that crates using it, such as this new SPARQL crate, would not end up importing the whole of sophia.

**Comment by Tpt** [comment]
@dwhitney Great! Thank you! I have done some changes in the parser that have significantly improve the speed of the current parser (migration to peg 0.6 and avoiding duplicate parsing).

**Comment by dwhitney** [comment]
Haven't had as much time as I'd like to look at this (still working from home). I found this parser. Have either of you taken a look at it? https://github.com/mattsse/nom-sparql

**Comment by Tpt** [comment]
@dwhitney I was not aware of this paper, thanks! There seems to be no code related to testing the parser against the W3C testsuite, I don't know how correct it is.

**Comment by dwhitney** [comment]
Yeah if your parser's performance has increased enough, perhaps there is no
need to invest the time in a new one, but before you made your
improvements, parsing was often the slowest part of the query by several
orders of magnitude.

**Comment by yever** [comment]
I was also thinking of implementing a SPARQL parser with [nom](https://github.com/Geal/nom) before I found https://github.com/mattsse/nom-sparql which was mentioned by @dwhitney.

It says it's a WIP in the README and there was no development for more than a year. I wonder whether @mattsse would be willing to adapt it to fit into `sophia`, allowing it, via abstraction, to reuse some types and code. If he only meant it as a WIP which he doesn't want to maintain, maybe he'd be willing to have it adapted and included in `sophia`.

**Comment by mattsse** [comment]
Hi there 👋

it's been a while since I've worked on it. It was a small side project I only hacked on for few weeks. Unfortunately I did not finish it so far that I was pleased with it and felt good about publishing it and moved on... 🙈

Currently I've got some time on my hands and if that crate could be useful for `sophia` I'd be willing to adapt/donate it. So long as `nom` as dependency is acceptable.

> There seems to be no code related to testing the parser against the W3C testsuite

I wasn't aware that there is a test suite, If you can point me to where i can find it, I'd be happy to test against it @Tpt 

fwiw the parser should be already feature complete(ish), so the most work would probably be 

* testing against the WC3 suite
* impl `Display` for every type, so that a roundtrip (string -> sparql -> string) is supported
* optional update to `nom` 0.6. (is still in alpha)

**Comment by MattesWhite** [comment]
I'm not the owner of `sophia` so take this with a grain of salt.

The aim of `sophia` is to provide a common API for RDF in Rust (#23), therefore, it is not intended to include a parser in `sophia` (the current implementations are more or less artefacts from before the split into several crates). A more fitting approach would be to develop a SPARQL API for `sophia`, i.e. a bunch of traits, base types and core functionality. So that third party crates, like `nom-sparql` can implement a parser against this API. In the end this should allow users to pick a parser that fits their needs best. In addition, this means that an implementation of a SPARQL engine is not required to include a parser.

---

@yever, @mattsse Nice to see new people working on `sophia` and its ecosystem :+1: 

@mattsse You can find out about the SPARQL test suite here: https://www.w3.org/2009/sparql/docs/tests/README.html

**Comment by Tpt** [comment]
@mattsse The recent versions of the test suite are here: https://github.com/w3c/rdf-tests/tree/gh-pages/sparql11
I use this repository as a git submodule in Oxigraph in order to be able to get quick feedbacks (<1s for the full SPARQL test suite). Here is my testsuite evaluation code: https://github.com/oxigraph/oxigraph/blob/master/testsuite/src/sparql_evaluator.rs
It contains also support of query and update evaluation tests.

Oxigraph already has `Display` implementations: https://github.com/oxigraph/oxigraph/blob/master/lib/src/sparql/algebra.rs via the Sparql* structs (the default display prints the algebra notation). During testing I check that the -> serialized -> parser returns the same tree.

**Comment by BonaBeavis** [comment]
I have written a [SPARQL parser](https://github.com/BonaBeavis/tree-sitter-sparql) with treesitter. Treesitter is very fast and has bindings for rust. Maybe this is of use.

**Comment by pchampin** [comment]
I have a very early implementation of a SPARQL engine for Sophia: https://github.com/pchampin/sophia_sparql

It should be integrated in v0.9 (but it might not be fully compliant by that time).
